inmoose/cohort_qc/cohort_metric.py

```diff
@@ -0,0 +1,864 @@
+        association_matrix = self._compute_pc_associations(pcs_after)
+
+        return (
+            association_matrix_before,
+            association_matrix,
+            pca_before,
+            pca_after,
+            pcs_before,
+            pcs_after,
+        )
+
+    def cohort_summary(self) -> dict:
+        """
+        Generate a summary of the cohort, including the number of samples, datasets, genes, and covariate combinations.
+
+        Returns
+        -------
+        dict
+            Summary of the cohort with details on samples, datasets, genes, and covariate combinations.
+        """
+        # Generate a summary of the cohort
+        num_samples = self.clinical_df.shape[0]
+        num_datasets = self.clinical_df[self.batch_column].nunique()
+        num_genes = self.data_expression_df.shape[0]
+        covariates_used = ", ".join(self.covariates)
+
+        # Calculate the number of samples by covariate combinations
+        covariate_combinations = (
+            self.clinical_df.groupby(self.covariates)
+            .size()
+            .reset_index(name="Number of Samples")
+        )
+
+        # Convert the covariate combinations and their counts to a readable format
+        covariate_combination_summary = covariate_combinations.to_dict(orient="records")
+
+        summary = {
+            "Number of Samples": num_samples,
+            "Number of Datasets": num_datasets,
+            "Number of Genes": num_genes,
+            "Covariates Used": covariates_used,
+            "Samples by Covariate Combination": covariate_combination_summary,
+        }
+
+        return summary
+
+    def median_absolute_deviation(self, data: pd.DataFrame) -> np.ndarray:
+        """
+        Compute the median absolute deviation (MAD) for each gene in the data.
+
+        Parameters
+        ----------
+        data : pd.DataFrame
+            Gene expression data frame.
+
+        Returns
+        -------
+        np.ndarray
+            Array of median absolute deviations for each gene.
+        """
+        return np.median(
+            np.abs(data.values - np.median(data.values, axis=1, keepdims=True)), axis=1
+        )
+
+    def quantify_correction_effect(self) -> tuple:
+        """
+        Compute a metric quantifying the effect of batch correction on the variability of gene expression data.
+
+        Returns
+        -------
+        tuple
+            A tuple containing the mean median absolute deviation before and after correction, and the correction effect metric.
+        """
+        # Compute MAD for after correction
+        mad_after = self.median_absolute_deviation(self.data_expression_df)
+
+        if self.data_expression_df_before is not None:
+            # Compute MAD for before correction if data is available
+            mad_before = self.median_absolute_deviation(self.data_expression_df_before)
+            effect_metric = np.mean(mad_after) / np.mean(mad_before)
```

Comment:
Do you have some documentation on this metric? I am interested to learn more

Reply:
It's a home made metric, there is no paper that uses the median absolute deviation metric to quantify the batch effect correction before/after in transcriptomics, but I had seen that it was useful instead of the standard deviation around the mean to detect outliers (some single cell pipelines use it to filter cells).
So I thought it might be useful to check that the batch effect correction had done its job by not adding variability/checking that variability/outliers had decreased with a ratio of MAD (a bit like the plot comparing variance explained by PCAs before and after correction). 
It's a very basic metric, but give a idea, and it was effective for the Kidney cohort with bad selection of covariate were the use of pycombat increase this metric and add variability to our data (but other metrics were also very bad)